### DISCUSSION TITLE
RGB color space not permitted on grayscale PNG

### DISCUSSION THREAD

**Discussion by juancarlangas** [ImageMagick/ImageMagick#2730] [Help] [answered]
Hi guys. Trying to add the "_checkerboard_" to my GIMP-**sRGBA** image:

![angel printed](https://user-images.githubusercontent.com/72962922/96223641-db3f6800-0f53-11eb-85ad-a1c848d44134.png)

with:
composite -compose Dst_Over -tile pattern:checkerboard original.png tiled.png

Magick generates the Warning Message: **"RGB color space not permitted on grayscale PNG"** and then produces a GRAY SCALE IMAGE.

Why does this occurs? Thats my unique sRGB black&alpha image with this issue. The others preserve the same GIMP sRGB profile even if they has only black and transparents pixels. 

I dont want to use the "-quiet" option, or even the "-strip" flag, but to know what does it happens and how to preserve the same profile in the resulting images. 

Thank you guys!

**Comment by t246246** [comment, marked as answer]
Imagemagick automatically convert RGB to Grayscale if image has only gray value.

-define colorspace:auto-grayscale=false prevent this behaviour, but it looks composite does not recognize it.  Probably you can do the same thing using convert.


Try to add -define png:color-type=6, or use PNG32: prefix.

% magick composite -compose Dst_Over -tile pattern:checkerboard 96223641-db3f6800-0f53-11eb-85ad-a1c848d44134.png -define png:color-type=6 out.png

or 

% magick composite -compose Dst_Over -tile pattern:checkerboard 96223641-db3f6800-0f53-11eb-85ad-a1c848d44134.png  PNG32:out.png

I also tried '-type TrueColorAlpha', but this does not solve that warning.

**Reply by juancarlangas** [reply]
@t246246 That works fine!! 💯 even if i dont know exactly what Im doing 🤣   I used the **PNG32:** prefix and the resulting image preserves the original GIMP profile 👍 

My only doubt is: this wasn't needed with another same-characteristics images. I attached an example for your consideration.

Thank you so much!!
![acueducto-plus printed](https://user-images.githubusercontent.com/72962922/96330976-0e9af900-101f-11eb-9b2a-f2239aa5a620.png)

**Reply by t246246** [reply]
I don't know the reason, but the checking the successful result file with identify, it returns bit-depth 4 for your 1st image (I name it sampleA.png) , whereas 8 for 2nd one, sampleB.png.

```
% magick composite -compose DstOver -tile pattern:checkerboard sampleA.png PNG32:outA.png
% magick composite -compose DstOver -tile pattern:checkerboard sampleB.png PNG32:outB.png
% magick identify -format '%[depth] %[bit-depth] %[colors]\n' outA.png outB.png
8 4 3
8 8 251
```
It implies imagemagick internally treat outA.png as 4 bit file after reading, probably because 1st result file has less color.

When your composite command is executed imagemagick wrote  grayscale png temporarily.

From debug output, 
```
% magick composite -verbose -debug all -compose DstOver -tile pattern:checkerboard sampleA.png outA.png 2>&1 | grep written
      PNG bit-depth written: 4
      PNG color-type written: 0

% magick composite -verbose -debug all -compose DstOver -tile pattern:checkerboard sampleB.png outB.png 2>&1 | grep written
      PNG bit-depth written: 8
      PNG color-type written: 3
```

color-type 0 is grayscale, so when it tries to apply icc profile later, png.c complains.

PNG32 (or even PNG00) setting force ImageMagick to writie the RGBA file.

```
% magick composite -verbose -debug all -compose DstOver -tile pattern:checkerboard sampleA.png PNG00:outA.png 2>&1 | grep written
      PNG bit-depth written: 8
      PNG color-type written: 6
```

**Reply by mikkorantalainen** [reply]
Isn't this actually a bug in ImageMagick? Shouldn't it apply ICC profile before checking if the image is actually a monocrome image or not? (Imagine color image that just happens to have all pixels with identical numerical values before applying ICC profile.)

**Comment by snibgo** [comment]
> Shouldn't it apply ICC profile before checking if the image is actually a monocrome image or not?

What do you mean by "apply"?

IM might read a colour image with an embedded RGB ICC profile. Then some processing removes the colour, so the image is grayscale. The ICC profile is still embedded. Then the image is saved to PNG, and this generates the warning.

I regard this as a minor problem in the PNG specifications and thus the processing. There is no logical problem with a RGB profile in a grayscale image. The profile specifies the RGB primaries which are not relevant to a grayscale image, and the transfer response curve which is relevant to a grayscale image.

So, I suggest you ignore the warning.

**Comment by fmw42** [comment]
If it is a warning, just add -quiet to your command line right after the convert or magick. That should cause the warning not to be listed to the terminal so you won't see it.
